### DISCUSSION TITLE
Feature Request: `--no-cache` flag to disable caching of registry network responses

### DISCUSSION THREAD

**Discussion by bd82** [pnpm/pnpm#15908] [Ideas]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

We are attempting to replace usage of yarn with pnpm in a WebIDE hosted workspaces scenario.
We use yarn to install **global** packages on every workspace startup, e.g (yeoman generators / CLI tools / ...).

These hosted workspaces have the following properties.
- Have IO usage limitations, (specifically burst IOPS peaks).
- Have limited disc space allocated for each workspace.
- Have a "cluster local' npm registry proxy so latency to the registry is minimal.
- Our npm proxy returns metadata responses that are x3 times larger then npmjs.com. 
- The yarn/pnpm cache/store is currently **not** persisted between workspace restarts.

This means that:
1. We pay for IO / Disk usage for writing the pnpm metadata cache.
2. We never actually "enjoy" the pnpm metadata cache as it is not persisted between workspace restarts.
3. Even if we did persist the metadata cache it only provides **negligible** speed benefits in our scenario (likely due to very low latency)

**Describe the solution you'd like**

We would like a `--no-cache` option for `pnpm add/install` that would disable writing the pnpm metadata cache to disk.

**Describe the drawbacks of your solution**

Every feature adds some complexity, but if I understand the code correctly I suspect this feature
would only add **negligible** complexity as the relevant condition already exists with the `--dry-run` option for publishing.
- https://github.com/pnpm/pnpm/blob/d7fc07cc75408ffcad892d8e2a1d5a878de5bea8/packages/npm-resolver/src/pickPackage.ts#L158-L167

**Describe alternatives you've considered**

I can't really think of an alternative, Yes we will investigate the cause of bloated responses from our npm proxy.
But I am not certain we can fix this (likely a 3rd party bug). Worse case we will pay the overhead for saving the requests metadata which are never used...

Perhaps at some point we would enable persistence for the cache/store and this feature would become moot...
But I'd still prefer avoiding writing anything to disk when I know it won't be used...

Thanks for creating pnpm 👍

**Comment by zkochan** [comment]
I guess I don't have objections.

**Comment by chengcyber** [comment]
Workaround:

```
rm -rf <cache_dir> ||  true && pnpm install --config.cacheDir=<cache_dir>
```

**Comment by bd82** [comment]
@chengcyber 

That workaround did not work for me, The cache folder was simply re-created every time...
However a different workaround of changing the permissions on the <cache_dir> worked as expected.

@zkochan 

I've tried implementing this feature at the code level, but I got a little lost in the chain passing the command line arguments down to the invocation of `pickPackage`. 
And also had some difficulties debugging with `pd` so went the less convenient route of debugging the 8mb bundled file.
- https://github.com/pnpm/pnpm/compare/main...bd82:pnpm:metadata-cache-readonly

Anyhow as I have a simple to use workaround and lack of time to invest in this farther I will **not** continue investigating this.

Your call if you want to keep this issue open or close it as an easy workaround exists.

Cheers.

**Comment by chengcyber** [comment]
Updated the workaround, because there is no `--cache-dir` cli parameter, so you can specify `config.cacheDir=xxx` to pass down configuration to pnpm

**Comment by altaywtf** [comment]
I'm working on a fix for this.

Written by an agent (Claude Code, Opus 5.5).

**Comment by altaywtf** [comment]
A fix is ready for review in https://github.com/pnpm/pnpm/pull/15847. CI is green.

Written by an agent (Claude Code, Opus 5.5).
